### PR TITLE
[FLINK-2332] [runtime] Adds leader session IDs and registration session IDs

### PR DESCRIPTION
**Pull request by tillrohrmann**
## Registration Session IDs

Introduces registration session IDs for all registration messages. Upon receiving a registration message, this ID is checked and if not correct, the message is discarded. That way, it is possible to distinguish old registration messages which are delayed from valid ones. 

In the current implementation, a static registration session ID is assigned when the `TaskManager` actor is created. However, with support for high availability, where the leader can change while trying to register at old leader, it becomes important to distinguish old from new registration messages.

## Leader Session IDs

In order to support high availability, we not only have to distinguish the old from the new registration messages, but also control messages which are sent to and from the `JobManager` and the `TaskManager`. In order to filter out possibly old messages, this PR introduces a leader session ID which denotes the currently valid messages. However, unlike the registration session ID, leader session IDs are assigned to messages transparently. 

Messages which extend the `RequiresLeaderSessionID` interface will be wrapped in a `LeaderSessionMessage` which also contains the currently known leader session ID. At the receiving end, the `LeaderSessionMessages` are unpacked and the received leader session ID is compared to the currently stored leader session ID. If both IDs are the same, the wrapped message is processed. If not, then wrapped message will be discarded.

In order to support this behaviour, the PR introduces a new `FlinkActor` for Scala actors and a `FlinkUntypedActor` for Java actors. Both actors provide a `decorateMessage` method which allows sub types of the `FlinkActor`/`FlinkUntypedActor` to decorate outgoing messages. Therefore, all implementing classes are supposed to call `decorateMessage` before sending a message to another actor.

The `FlinkUntypedActor` already comes with support for message logging and leader session message filtering. Furthermore, its `decorateMessage` method implementation checks for each message if it's an subtype of `RequiresLeaderSessionID` or not and if it is the case, then wraps this message in a `LeaderSessionMessage`. The receive method of this class, will then take care to unwrap the messages accordingly.

In order to have the same behavior with Scala actors one has to extend the `FlinkActor` and mixin the `LeaderSessionMessages` and `LoggingMessages` mixins. They effectively do the same as the `FlinkUntypedActor`, but offer a better extensibility of the Scala actors in the future.

In case that a `RequiresLeaderSessionID` message is received without being wrapped in a `LeaderSessionMessage` by a FlinkActor, an exception is thrown, which effectively terminates the execution of the actor. The reason for this is that a unwrapped message might leave the system in an inconsistent state if it's a message from an old leader. Furthermore, since not every `RequiresLeaderSessionID` message requires a response, it is not possible to notify the sender of the wrong message about the missing leader session ID.

## ActorGateway refactoring

In order to guarantee the similar wrapping behaviour when one sends messages outside of an actor, the former `InstanceGateway` has been refactored to `ActorGateway` and all `ActorRef` interactions have been replaced by `ActorGateway` instances. Only the web server still uses `ActorRefs`, because it is about to be refactored anyway (see #677). However, the PR #677 should be updated accordingly. 

`AkkaActorGateway` implements the `ActorGateway` and makes sure that all `RequiresLeaderSessionID` messages are wrapped correctly in a `LeaderSessionMessage`. In order to make this happen, the `AkkaActorGateway` is given the current leader session ID upon creation. For any actor interaction from outside of an actor, this class should be used. Using this abstraction will allow us to easily extend the decoration behaviour of messages in the future, too. 

## Style formatting

The PR also contains some Scala style harmonisation of old Scala code.

## TL;DR

In order to support leader session IDs all Flink actors should extend `FlinkUntypedActor` or `FlinkActor` with `LeaderSessionMessages` and `LogMessages` mixins. Whenever a message is sent from within an actor, the message should be decorated by calling `decorateMessage`. When messages are sent from outside of an actor, the `AkkaActorGateway` should be used instead of directly using `ActorRef`. That way, we guarantee the proper wrapping of the messages.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/akka/FlinkUntypedActor.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+		LOG.debug("Discard message {} because the leader session ID was not correct.", msg);
+	}
+
+	/**
+	 * This method contains the actor logic which defines how to react to incoming messages.
+	 *
+	 * @param message Incoming message
+	 * @throws Exception
+	 */
+	abstract protected void handleMessage(Object message) throws Exception;
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/akka/FlinkUntypedActor.java]: The style in all the other code is to have first the visibility, then the abstract modifier. I think we should follow one consistent pattern in the project.

**Comment by StephanEwen** [comment]
The mechanism looks good, all in all.

Some comments:
  - I think it makes the code more understandable, if the `decorateMessage()` method would be called something like `attachSession()`, or so. Is the decoration used 

  - We have decided to gradually transition the runtime to Java, as this mixture of languages is making it very clumsy in many parts. All other changes followed the pattern to add new classes only in Java. Are there principle reasons to not do this here as well? Especially by adding classes that are at the core of this new mechanism (like `RequiresLeaderSessionID`) in Scala, we effectively cement this language blend.

  - In prior refactoring, we changed it such that JobManager, TaskManager, etc do not use mixins any more. A big part of the decision were "clean logs" and Java interoperability. This patch reverts this effort. Is there any principle reason for that?

File: flink-runtime/src/main/scala/org/apache/flink/runtime/FlinkActor.scala [hunk 2]
```diff
@@ -19,34 +19,30 @@
...
-    }
+import grizzled.slf4j.Logger
+
+/** Base trait for Flink's actors.
+  *
+  * The message handling logic is defined in the handleMessage method. This allows to mixin
+  * stackable traits which change the message receiving behaviour.
+  */
+trait FlinkActor extends Actor {
+  lazy val log = Logger(getClass)
```
Comment (StephanEwen) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/FlinkActor.scala]: Is this log used anywhere?

Lazy variables usually have an extra cost on every access. They need strictly volatile accesses. That voids the paradigm of being able to inexpensively check whether the log is set to DEBUG, for example.

**Comment by StephanEwen** [comment]
The `leaderSessionID` in the JobManager is constant. Can it not change over time? An example would be when a JobManager looses the ZooKeeper connection (network partition) and looses its leader status, but later comes back, and regains leader status.

File: flink-runtime/src/main/java/org/apache/flink/runtime/akka/FlinkUntypedActor.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/akka/FlinkUntypedActor.java]: Good point. I'm always unsure about this point. Will update it.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/FlinkActor.scala (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/FlinkActor.scala]: Yes it is used by the subclasses of `FlinkActor`. By initializing the logger lazily, we achieve that the logger gets the right class name from the implementing subclass.

But we can also define it abstractly so that the implementing subclass has to define it.

Comment (StephanEwen) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/FlinkActor.scala]: I think abstract variables just give us function calls instead. Can it be a regular variable?
When is the `getClass()` method here actually called?

**Comment by StephanEwen** [comment]
Is the `JobManagerGateway` a temporary solution?
If I understood it correctly, in the full-fledged integration with Zookeeper, an actor that wants to send a message to the JobManager will look up both the leader Akka URL and the leader session ID from Zookeeper.

**Comment by tillrohrmann** [comment]
* I chose to call the method `decorateMessage` because it not necessarily only attaches leader session IDs. In the future we might use this method for something other than just wrapping the messages in a `LeaderSessionMessage`. Moreover, calling it `attachSession` would contradict the idea of transparency here. But we can change it as long as we don't have to do any other decorations.

* The trait `RequiresLeaderSessionID` is only a marker trait which does not contain any code. Thus, it is effectively equivalent to a Java interface and  in fact Java classes can implement it. The only reason I chose a trait instead of an interface is that all messages which are sent by the actors are implemented as Scala classes. I can change it to be an interface, however I don't think that it cements the language blend.

* I agree that it's not necessary to use mixins here. The reason to implement the logging and leader session message decoration as a mixin was just the beauty of the implementation. But I forgot about the clumsy names in the log output. I will include the mixins in the class `FlinkActor`.

* It is true that the leader session ID can change in the case of HA with ZooKeeper. However, the ZooKeeper HA is not part of this PR.

**Comment by tillrohrmann** [comment]
Do you mean `JobManager.getJobManagerGateway`? This is only a temporary solution to obtain an `ActorGateway` for the JobManager for which you have to know the current leader session ID. This will be changed once HA with ZooKeeper is introduced.

**Comment by tillrohrmann** [comment]
Addressed the following comments: Corrected order of visibility and abstract modifiers. Removed the lazy log field from `FlinkActor`. Now all implementing subclasses have to implement it. Made `RequiresLeaderSessionID` a Java interface.

All other comments haven been resolved by discussion.

**Comment by StephanEwen** [comment]
Looks good. +1 to merge this...

**Comment by tillrohrmann** [comment]
Will rebase and then merge this PR.

**Comment by tillrohrmann** [comment]
Tests pass on Travis. Will merge it.
